ember-server/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala

```diff
@@ -151,18 +152,78 @@ private[server] object ServerHelpers {
         }
         .drain
 
-    sg.server[F](bindAddress, additionalSocketOptions = additionalSocketOptions)
+    val handler = sg.server[F](bindAddress, additionalSocketOptions = additionalSocketOptions)
       .interruptWhen(shutdown.signal.attempt)
-      // Divorce the scopes of the server stream and handler streams so the
-      // former can be terminated while handlers complete.
-      .prefetch
       .map { connect =>
         shutdown.trackConnection >>
           Stream
             .resource(connect.flatMap(upgradeSocket(_, tlsInfoOpt)))
             .flatMap(withUpgradedSocket(_))
       }
-      .parJoin(maxConcurrency)
-      .drain
+      
+    forking(handler)
+  }
+
+  // Similar to parJoin with a few semantic differences:
+  // - Inner stream values are not emitted to the new stream
+  // - The outer stream can terminate and finalize early
+  def forking[F[_], O](streams: Stream[F, Stream[F, O]])(implicit F: Concurrent[F]): Stream[F, INothing] = {
+    val fstream = for {
+      done <- SignallingRef[F, Option[Option[Throwable]]](None)
+      running <- SignallingRef[F, Long](1)
+    } yield {
+      val incrementRunning: F[Unit] = running.update(_ + 1)
+      val decrementRunning: F[Unit] = running.update(_ - 1)
+      val awaitWhileRunning: F[Unit] = running.discrete.dropWhile(_ > 0).take(1).compile.drain
+
+      val stop: F[Unit] =
+        done.update {
+          case None => Some(None)
+          case x => x
+        }
+
+      val stopSignal: Signal[F, Boolean] =
+        done.map(_.nonEmpty)
+
+      def handleResult(result: Either[Throwable, Unit]): F[Unit] =
+        result match {
+          case Right(_) => F.unit
+          case Left(err) => 
+            done.update {
+              case None => Some(Some(err))
+              case x => x
+            }
+        }
+
+      def runInner(inner: Stream[F, O]): F[Unit] =
+        incrementRunning >> 
+          inner
+            .interruptWhen(stopSignal)
+            .compile
+            .drain
+            .attempt
+            .flatMap(handleResult) >> decrementRunning
+
+      val runOuter: F[Unit] =
+        streams
+          .evalMap(inner => F.start(runInner(inner)))
```

Comment:
`parJoin` messes around with scope leases a bit, which is what I'm assuming prevents finalizers for the outer stream to be run until after all inner streams close.

Reply:
yes, this is what the leasing does